### PR TITLE
feat(template): Added pybind11 as an option for build system 

### LINKED ISSUE #70: Add support for pybind11 build-system (setuptools + pybind11)

### PR DESCRIPTION
**Pull request by ayeankit**
## Pull Request description

Fixes #70 

1. Activate option in TUI to pybind11
2. Adding documentation about pybind11
3. Adding pybind11 as a build-system:
4. Creating a pybind11-pyproject.toml
5. Created cmake file for pybind11.
6. Editing post_gen_project.py
7. Creating a smoke test (build-system.sh)
8. Editing cookicutter.json
9. Added pybind11 in Readme.md

## How to test these changes

* ```...```

## Pull Request checklists

This PR is a:
- [x] bug-fix
- [x] new feature
- [ ] maintenance

About this PR:
- [ ] it includes tests.
- [ ] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [ ] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:
- [ ] I have reviewed the changes and it contains no misspelling.
- [ ] The code is well commented, especially in the parts that contain more complexity.
- [ ] New and old tests passed locally.

## Reviewer's checklist

Copy and paste this template for your review's note:

```
## Reviewer's Checklist

- [ ] I managed to reproduce the problem locally from the `main` branch
- [ ] I managed to test the new changes locally
- [ ] I confirm that the issues mentioned were fixed/resolved .
```

### CODE REVIEW

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/skcdemo.cpp
Comment (xmnlab) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/skcdemo.cpp:7]: maybe you should change the name of the file to something more generic than `skcdemo`
maybe example.cpp ... or as it uses pybind11 .. maybe pybind11.cpp ..

File: tests/smoke/base.sh
Comment (xmnlab) [inline tests/smoke/base.sh:60]: ideally you should run just `pip install -e .` and it should run cmake under the hook

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11-pyproject.toml
Comment (xmnlab) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11-pyproject.toml:119]: I don't think it needs hatch session here

**Review by xmnlab — CHANGES_REQUESTED** [review]
I am not super familiar with pybind11
but maybe it would be necessary to have an extra setup.py file
ref: https://github.com/pybind/python_example/blob/master/setup.py

File: docs/guide.md
Comment (Saransh-cpp) [inline docs/guide.md:338]: Should we call it something like `setuptools + pybind11` to be more clear, given that `pybind11` itself is not a backend?

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/CMakeLists.txt
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/CMakeLists.txt:23]:
```suggestion
{%- endif %}

```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11.cpp
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11.cpp:9]:
```suggestion
}

```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py:16]: All of this is already specified in `pyproject.toml`. We would need something like this - https://github.com/scientific-python/cookie/blob/main/%7B%7Bcookiecutter.project_name%7D%7D/%7B%25%20if%20cookiecutter.backend%3D%3D'pybind11'%20%25%7Dsetup.py%7B%25%20endif%20%25%7D

File: docs/guide.md
Comment (Saransh-cpp) [reply docs/guide.md:338]: Or atleast using "setuptools + pybind11" in all the user-facing locations, and just "pybind11" in the internals of `scicookie`.

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Great work, thanks, @ayeankit! Some comments below -

File: docs/guide.md
Comment (xmnlab) [reply docs/guide.md:338]: > Or atleast using "setuptools + pybind11" in all the user-facing locations, and just "pybind11" in the internals of `scicookie`.

that makes sense to me.

Comment (ayeankit) [reply docs/guide.md:338]: Wouldn't that be confusing since we are using setuptools as default build-sytem?

Comment (xmnlab) [reply docs/guide.md:338]: not really ... setuptools is the default option because it was the first one historically, but a lot of people want to kill that. That is why a bunch of project are looking for new alternatives.
but for c/c++ code setuptools is still a great option. 
so specifying setuptools + pybind11 makes a lot of sense, because it is clear about its structure.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/CMakeLists.txt
Comment (xmnlab) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/CMakeLists.txt:1]: where are you declaring this variable? `SKBUILD_PROJECT_NAME`

Comment (xmnlab) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/CMakeLists.txt:1]: it doesn't look correct to resuse the name sk... for both build system
also the first two lines seem to be duplicated .. probably you could move it to some first lines in order to reuse that for both build system

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11.cpp
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11.cpp:5]: `skcdemo` can be called maybe `pybinddemo`?

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]

File: src/scicookie/{{cookiecutter.project_slug}}/Makefile
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/Makefile:119]: Are you sure this command exists?

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/CMakeLists.txt
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/CMakeLists.txt:17]: I think the `src` should not be here? Or it should be `src/<package_name>/main.cpp`?

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/pybind11-pyproject.toml:2]: IIRC, editable install with `pyproject.toml` support was first supported with setuptools version `65`. Could you please check and confirm?

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py:1]: We are not adding type annotations anywhere in this file. What purpose is this import serving then?

File: tests/smoke/base.sh
Comment (Saransh-cpp) [inline tests/smoke/base.sh:60]: Yes! This `-e` would not work with `pyproject.toml` and `setuptools<65`, IIRC.

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Some comments below that might push you in the right direction for fixing the CI error 🙂

File: src/scicookie/{{cookiecutter.project_slug}}/Makefile
Comment (xmnlab) [reply src/scicookie/{{cookiecutter.project_slug}}/Makefile:119]: I think you can use something like `python3 -m build`  https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#packaging-your-project

ensure you have `build` as a dependency in pyproject.toml/build-system

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py:13]: You should take a look at the path here again.

File: tests/smoke/base.sh
Comment (Saransh-cpp) [inline tests/smoke/base.sh:60]: This should still be `-e`

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]

File: src/scicookie/{{cookiecutter.project_slug}}/Makefile
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/Makefile:119]:
```suggestion
	python -m build
```
and add `build` as a dependency in `build-system.requires` in `pyproject.toml`.

File: src/scicookie/hooks/post_gen_project.py
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:311]: I think `PROJECT_DIRECTORY` automatically checks for flat or `src` layout?

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]

File: src/scicookie/hooks/post_gen_project.py
Comment (ayeankit) [reply src/scicookie/hooks/post_gen_project.py:311]: I think it for `main.cpp` which should be in `src` directory like this. 
your_project/
|-- setup.py
|-- pyproject.toml
|-- src/
|   |-- main.cpp
|   |-- other_cpp_files.cpp
|   |-- your_module_name_here.cpp
|   |-- your_module_name_here.h
|-- your_package/
|   |-- __init__.py
|   |-- your_module_name_here.py
|-- tests/
|   |-- test_your_module_name_here.py
|-- README.md
|-- LICENSE

**Comment by ayeankit** [comment]
Hey @xmnlab , All test cases passed !!, I have changed `pybind11` to `setuptools + pybind11` in user-facing locations as suggested by Saransh. Can you please review PR?

File: src/scicookie/hooks/post_gen_project.py
Comment (xmnlab) [reply src/scicookie/hooks/post_gen_project.py:311]: this is an interesting discussion ... maybe we can return to this discussion to check how we should proceed not just for this build-system but maybe for all build-systems.
also maybe in our repo, maybe we could have this files in folders called by the programming language (`cpp`, `rust`, etc)

File: tests/smoke/base.sh
Comment (xmnlab) [reply tests/smoke/base.sh:60]: in my last PR I removed `-e` from everywhere, so we will be sure to test the package correctly (as much as possible). my main concern is about testing this extension files .. so we would ensure to test if that is in the package .. maybe the test is not good enough yet ... but at least it is a first step

File: tests/smoke/base.sh
Comment (xmnlab) [inline tests/smoke/base.sh:57]:
```suggestion
  pip install .
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py
Comment (xmnlab) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/setup.py:13]: this is an interesting discussion .. but I think we can have this discussion after we merge this PR
I am also checking this discussion with Leah (pyOpenSci) to see if there is already something recommended by pyOpenSci

**Comment by ayeankit** [comment]
hey @xmnlab , I did the required changes, and added `"build>=0.10.0"`, in dependency in `pyproject.toml` for pybind11. Can you please review it again?

**Review by xmnlab — APPROVED** [review]
thanks for working on that @ayeankit !

lgtm! 

the only thing that is not working well is the version, as you can see in the logs:

```
creating osl-python-package-0.0.0
creating osl-python-package-0.0.0/src
creating osl-python-package-0.0.0/src/osl_python_package
creating osl-python-package-0.0.0/src/osl_python_package.egg-info
creating osl-python-package-0.0.0/tests
copying files to osl-python-package-0.0.0...
copying LICENSE -> osl-python-package-0.0.0
copying README.md -> osl-python-package-0.0.0
copying pyproject.toml -> osl-python-package-0.0.0
copying setup.py -> osl-python-package-0.0.0
copying src/main.cpp -> osl-python-package-0.0.0/src
copying src/osl_python_package/__init__.py -> osl-python-package-0.0.0/src/osl_python_package
copying src/osl_python_package/osl_python_package.py -> osl-python-package-0.0.0/src/osl_python_package
copying src/osl_python_package.egg-info/PKG-INFO -> osl-python-package-0.0.0/src/osl_python_package.egg-info
copying src/osl_python_package.egg-info/SOURCES.txt -> osl-python-package-0.0.0/src/osl_python_package.egg-info
copying src/osl_python_package.egg-info/dependency_links.txt -> osl-python-package-0.0.0/src/osl_python_package.egg-info
copying src/osl_python_package.egg-info/requires.txt -> osl-python-package-0.0.0/src/osl_python_package.egg-info
copying src/osl_python_package.egg-info/top_level.txt -> osl-python-package-0.0.0/src/osl_python_package.egg-info
copying tests/test_osl_python_package.py -> osl-python-package-0.0.0/tests
```

but we can address this issue in a follow-up.

**Pull request merged**
